invoiceregistry_test.go

```diff
@@ -299,6 +299,106 @@ func TestCancel(t *testing.T) {
 	cancelUpdates()
 }
 
+func TestOverpayment(t *testing.T) {
+	defer test.Timeout()()
+
+	c := newRegistryTestContext(t, true)
+
+	// Add invoice.
+	invoice, preimage := c.createInvoice(1, time.Hour)
+	rHash := preimage.Hash()
+
+	// Subscribe to updates for invoice.
+	updateChan, cancelUpdates := c.subscribe(preimage.Hash())
+	acceptChan, cancelAcceptUpdates := c.subscribeAccept()
+
+	resolved := make(chan HtlcResolution, 3)
+
+	notifyHtlc := func(id uint64, amt int64) {
+		c.registry.NotifyExitHopHtlc(&registryHtlc{
+			rHash:         rHash,
+			amtPaid:       lnwire.MilliSatoshi(amt),
+			expiry:        100,
+			currentHeight: 0,
+			circuitKey: types.CircuitKey{
+				HtlcID: id,
+			},
+			resolve: func(r HtlcResolution) {
+				resolved <- r
+			},
+			payload: &testPayload{
+				amt:     lnwire.MilliSatoshi(c.testAmt),
+				payAddr: invoice.PaymentAddr,
+			},
+		})
+	}
+
+	// Test with full overpayment in a single HTLC.
+	notifyHtlc(1, c.testAmt+1)
+
+	resolution := <-resolved
+
+	require.IsType(t, &HtlcFailResolution{}, resolution)
+	require.Equal(t, ResultHtlcSetOverpayment, resolution.(*HtlcFailResolution).Outcome)
+
+	// We should have neither acceptance nor update here.
+	select {
+	case <-acceptChan:
+		require.FailNow(t, "Invoice accepted incorrectly")
+	case <-updateChan:
+		require.FailNow(t, "Update generated incorrectly")
+	default:
+	}
+
+	// Test with half of the amount in one HTLC and a slight overpayment
+	// in a second HTLC.
+	first := c.testAmt / 2
+	rest := c.testAmt - first
+
+	notifyHtlc(2, first)
+	notifyHtlc(3, rest+1000)
+
+	resolution = <-resolved
+	require.IsType(t, &HtlcFailResolution{}, resolution)
+	require.Equal(t, ResultHtlcSetOverpayment, resolution.(*HtlcFailResolution).Outcome)
+
+	// We should not have a second resolution, an update, or an accept here.
+	select {
+	case <-acceptChan:
+		require.FailNow(t, "Invoice accepted incorrectly")
+	case <-updateChan:
+		require.FailNow(t, "Update generated incorrectly")
+	case <-resolved:
```

Comment:
It's expected to be empty to ensure only one HTLC (ID 3) is resolved as checked above, and that there are no other resolved HTLCs. I could add a `<-time.After(...)` here as well instead of default, to make sure.

Reply:
As above, changed to 100ms timeout.